### ISSUE TITLE
[feature request] support ContextVars

### ISSUE THREAD

**Issue by dimaqq** [florimondmanca/asgi-lifespan#45]
It seem that setting context variables in `on_startup` is not supported.

My naïve understanding is that there needs to be a new (copied) context in which all of `on_startup`, asgi request and then `on_shutdown` are ran.

Otherwise, the value that's been set is not accessible in the request handlers and/or the context var cannot be reset in the shutdown function using the token that was generated in the startup function.

**Comment by florimondmanca** [comment]
Hi,

Would this be a ASGI server/framework-dependant issue instead? Or do you mean that a no-server, plain asyncio code example produces this behavior as well?

Asking because I _think_ I saw a similar discussion over either Uvicorn or Starlette repos — mostly likely Starlette, which would run startup handlers in a long running lifespan task, and handle requests in another task, leading to different contexts and more broadly a lack of contextvars support for lifespan in Starlette

**Comment by dimaqq** [comment]
@florimondmanca I think you may be right, I'm using fastapi/starlette/uvicorn and andeed my context var setup needs tricky context saving/switching because lifespan is a separate call (from middleware's point of view).

**Comment by florimondmanca** [comment]
Right. I _think_ that’s the current state of things, so gently going to close this as being not related to ASGI-lifespan per se. Thanks :)

**Comment by dimaqq** [comment]
I think I know what happened...

I made a middelware that listens to lifespan events, and I'm using with fastapi/starlette/uvicorn.
It works fine in that config.

However, in the tests, I'm trying to use the context manager from this project.
And that does not work for me.

Key difference is that I `await` something in my lifespan hook.

| # | step | fastapi  | asgi-lifespan |
| ------------- | ------------- | ------------- | ------------- |
| Content Cell  | Content Cell  | Content Cell  | Content Cell  |
| Content Cell  | Content Cell  | Content Cell  | Content Cell  |
| Content Cell  | Content Cell  | Content Cell  | Content Cell  |
| Content Cell  | Content Cell  | Content Cell  | Content Cell  |
| Content Cell  | Content Cell  | Content Cell  | Content Cell  |

**Comment by florimondmanca** [comment]
> This library kindof assumes that middleware lifespan processing is synchronous.

I'm not sure this is accurate, as we _do_ `await` for an event that's set once we receive `lifespan.startup.complete`, right?

https://github.com/florimondmanca/asgi-lifespan/blob/fbb0f440337314be97acaae1a3c0c7a2ec8298dd/src/asgi_lifespan/_manager.py#L29-L33

I think at this point sharing a bit of minimal reproduction code would be super helpful to see if there's anything we should be doing differently.

**Comment by dimaqq** [comment]
🤔  the code appears to do the right thing...
starlette lifecycle is indeed quite peculiar, and as @ojii pointed out (separately) my middleware may not be doing it right.
another possibility is that there's a subtle bug somewhere

**Comment by jpgxs** [comment]
Hi, I've also been running into this issue 😢

I haven't dug into the code yet but it seems that startup and shutdown are being invoked in different contexts.

With this minimal example:

```python
from fastapi import FastAPI
from contextvars import ContextVar

api = FastAPI()
ctx_var = ContextVar('mycontextvar')

async def lifespan(*args):
    token = ctx_var.set(True)
    try:
        yield
    finally:
        ctx_var.reset(token)

api.router.lifespan_context = lifespan
```

The below exception occurs:
```
unhandled exception during asyncio.run() shutdown
task: <Task finished name='Task-6' coro=<<async_generator_athrow without __name__>()> exception=ValueError("<Token var=<ContextVar name='mycontextvar' at 0x7fe40ec69e50> at 0x7fe40ec78680> was created in a different Context")>
Traceback (most recent call last):
  File "api.py", line 73, in lifespan
    yield
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "api.py", line 75, in lifespan
    ctx_var.reset(token)
ValueError: <Token var=<ContextVar name='mycontextvar' at 0x7fe40ec69e50> at 0x7fe40ec78680> was created in a different Context
```

**Comment by florimondmanca** [comment]
A few questions to get the investigation going…

- Is there an even simpler example to allow us to reproduce this? Does it reproduce without Starlette? (Manual ASGI lifespan app with this async context manager lifespan handler)
- Does it reproduce on other ASGI servers? Hypercorn?

If we know in what happens in these situations then we may be able to pinpoint where the problem might be.

Now that I think about it, I said at the beginning of this thread that lifespan is a long running task, but more precisely I believe that’s how Uvicorn does it. Still, this « token was created in a different Context » error seems different, even odd. I looked up Starlette code, and the lifespan_context there runs as a single async with call, no strange task stuff going on. Which strengthens my belief that we might want to look at how Uvicorn runs lifespan..

**Comment by florimondmanca** [comment]
It seems like [ASGI lifespan state] allows addressing this use case, by allowing to store state initialized in the lifespan task for use in other places in the app.

We just merged #57 which adds support for it in this package, so I'll close this now!
